### PR TITLE
snap-bootstrap: support auto-detect device in create-partitions

### PR DESCRIPTION
**Pull request by mvo5**
The current code in `snap-bootstrap create-partitions` expects
that we pass a device. This is not really needed because the
device to use can be figured out from the gadget-dir. I.e. it's
enough to find the "system-seed" partition and get the device
from that.

This means the layers above partition creation do not need to
be in the business of finding the right device and all low-level
handling is in the same (code) area.

If this looks reasonable I can expand tests and it should make
#7762 a lot easier to review (because 7762 essentially just
needs to run a single binary/call a single function).

### CODE REVIEW

File: cmd/snap-bootstrap/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -29,19 +29,40 @@ type Options struct {
 	// will contain encryption later
 }
 
+func deviceFromRole(lv *gadget.LaidOutVolume, role string) (device string, err error) {
+	for _, vs := range lv.LaidOutStructure {
+		if vs.Role == role {
```
Comment (pedronis) [inline cmd/snap-bootstrap/bootstrap/bootstrap.go]: this part of the finding maybe should be a method on gadget.*Volume

File: cmd/snap-bootstrap/bootstrap/bootstrap.go [hunk 2]
```diff
@@ -29,19 +29,40 @@ type Options struct {
...
+			device, err = gadget.FindDeviceForStructure(&vs)
+			if err != nil {
+				return "", err
+			}
+			return gadget.DiskFromPartition(device)
+		}
+	}
+	return "", fmt.Errorf("cannto find role %s in gadget", role)
```
Comment (cmatsuoka) [inline cmd/snap-bootstrap/bootstrap/bootstrap.go]: s/cannto/cannot/

File: cmd/snap-bootstrap/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -29,19 +29,40 @@ type Options struct {
...
-		return fmt.Errorf("cannot use empty device node")
-	}
 
 	lv, err := gadget.PositionedVolumeFromGadget(gadgetRoot)
 	if err != nil {
 		return fmt.Errorf("cannot layout the volume: %v", err)
 	}
 
+	// XXX: the only situation where auto-detect is not desired is
+	//      in (spread) testing - consider to remove forcing a device
```
Comment (cmatsuoka) [inline cmd/snap-bootstrap/bootstrap/bootstrap.go]: Won't `gadget.DiskFromPartition()` work also for loop devices, making it usable in spread tests?

File: cmd/snap-bootstrap/bootstrap/bootstrap.go (hunk 2 above)
Comment (mvo5) [reply cmd/snap-bootstrap/bootstrap/bootstrap.go]: Thanks, fixed!

File: cmd/snap-bootstrap/bootstrap/bootstrap.go (hunk 3 above)
Comment (mvo5) [reply cmd/snap-bootstrap/bootstrap/bootstrap.go]: Yes, I will look into adding a spread test for this.

File: cmd/snap-bootstrap/bootstrap/bootstrap_test.go [hunk 4]
```diff
@@ -209,3 +214,70 @@ func layoutFromYaml(c *C, gadgetYaml, volume string) *gadget.LaidOutVolume {
 	c.Assert(err, IsNil)
 	return pv
 }
+
+const mockUc20GadgetYaml = `volumes:
```
Comment (pedronis) [inline cmd/snap-bootstrap/bootstrap/bootstrap_test.go]: s/Uc/UC/ please

File: cmd/snap-bootstrap/bootstrap/bootstrap.go [hunk 5]
```diff
@@ -29,19 +29,42 @@ type Options struct {
 	// will contain encryption later
 }
 
+func deviceFromRole(lv *gadget.LaidOutVolume, role string) (device string, err error) {
+	for _, vs := range lv.LaidOutStructure {
+		// XXX: this part of the finding maybe should be a
+		// method on gadget.*Volume
+		if vs.Role == role {
```
Comment (zyga) [inline cmd/snap-bootstrap/bootstrap/bootstrap.go]: Being somewhat sleepy still, we validate that there's only one instance of each role, right?

File: cmd/snap-bootstrap/bootstrap/bootstrap_test.go [hunk 6]
```diff
@@ -28,25 +28,30 @@ import (
...
 
 var _ = Suite(&bootstrapSuite{})
 
 // XXX: write a very high level integration like test here that
 // mocks the world (sfdisk,lsblk,mkfs,...)? probably silly as
 // each part inside bootstrap is tested and we have a spread test
 
+func (s *bootstrapSuite) SetUpTest(c *C) {
+	s.dir = c.MkDir()
+	dirs.SetRootDir(s.dir)
```
Comment (zyga) [inline cmd/snap-bootstrap/bootstrap/bootstrap_test.go]: Missing restore?

File: cmd/snap-bootstrap/bootstrap/bootstrap.go (hunk 5 above)
Comment (mvo5) [reply cmd/snap-bootstrap/bootstrap/bootstrap.go]: We do have a check for this but I am not sure we actually have a unittest for this, I will add one.

**Pull request merged**